x/confidentialtransfers/types/apply_pending_balance.go

```diff
@@ -0,0 +1,119 @@
+	"errors"
+	"strconv"
+
+	"github.com/sei-protocol/sei-chain/x/confidentialtransfers/utils"
+	"github.com/sei-protocol/sei-cryptography/pkg/encryption"
+	"github.com/sei-protocol/sei-cryptography/pkg/encryption/elgamal"
+)
+
+// MsgApplyPendingBalance is a message to apply the pending balance to the available balance
+type ApplyPendingBalance struct {
+	Address                        string
+	Denom                          string
+	NewDecryptableAvailableBalance string
+	CurrentPendingBalanceCounter   uint32
+	CurrentAvailableBalance        *elgamal.Ciphertext
+}
+
+// NewApplyPendingBalance creates a new MsgApplyPendingBalance instance
+func NewApplyPendingBalance(
+	privKey ecdsa.PrivateKey,
+	address, denom,
+	currentDecryptableBalance string,
+	currentPendingBalanceCounter uint16,
+	currentAvailableBalance,
+	currentPendingBalanceLo,
+	currentPendingBalanceHi *elgamal.Ciphertext) (*ApplyPendingBalance, error) {
+	aesKey, err := encryption.GetAESKey(privKey, denom)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the current balance from the decryptable balance.
+	currentBalance, err := encryption.DecryptAESGCM(currentDecryptableBalance, aesKey)
+	if err != nil {
+		return nil, err
+	}
+
+	teg := elgamal.NewTwistedElgamal()
+	keyPair, err := teg.KeyGen(privKey, denom)
+	if err != nil {
+		return nil, err
+	}
+
+	// Calculate the pending balances that we need to add to the available balance.
+	loBalance, err := teg.Decrypt(keyPair.PrivateKey, currentPendingBalanceLo, elgamal.MaxBits32)
+	if err != nil {
+		return nil, err
+	}
+
+	hiBalance, err := teg.DecryptLargeNumber(keyPair.PrivateKey, currentPendingBalanceHi, elgamal.MaxBits48)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the pending balance by combining the lo and hi bits
+	pendingBalance := utils.CombineTransferAmount(uint16(loBalance), uint32(hiBalance))
+
+	// Sum the balances to get the new available balance
+	newDecryptedAvailableBalance := currentBalance + pendingBalance
+
+	// Check for overflow: if the sum is less than one of the operands, an overflow has occurred.
+	if newDecryptedAvailableBalance < currentBalance {
+		return nil, errors.New("addition overflow: total balance exceeds uint64")
+	}
+
+	// Encrypt the new available balance
+	newDecryptableAvailableBalance, err := encryption.EncryptAESGCM(newDecryptedAvailableBalance, aesKey)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ApplyPendingBalance{
+		Address:                        address,
+		Denom:                          denom,
+		NewDecryptableAvailableBalance: newDecryptableAvailableBalance,
+		CurrentPendingBalanceCounter:   uint32(currentPendingBalanceCounter),
+		CurrentAvailableBalance:        currentAvailableBalance,
+	}, nil
+}
+
+// ApplyPendingBalanceDecrypted is the decrypted version of ApplyPendingBalance
+func (r *ApplyPendingBalance) Decrypt(decryptor *elgamal.TwistedElGamal, privKey ecdsa.PrivateKey, decryptAvailableBalance bool, address string) (*ApplyPendingBalanceDecrypted, error) {
```

Comment:
we may need to add inputs validation for decryptor, privkey

Reply:
Added validation for decryptor to prevent panic, but not for privkey since empty privKey will be reported as an invalid privKey.